Fixed issue 1003: Show Log does not remember the from-date

YueLinHo opened the pull request:
TortoiseGit [issue 1003](https://code.google.com/p/tortoisegit/issues/detail?id=1003)
([issue 1164](https://code.google.com/p/tortoisegit/issues/detail?id=1164), [issue 2078](https://code.google.com/p/tortoisegit/issues/detail?id=2078))

YueLinHo commented:
About last commit:
Can right click on FROM date control of Log Message dialog.

AndrewLane commented:
Thanks for this fix.

YueLinHo commented:
@AndrewLane Thanks for your feedback. ^_^

YueLinHo commented on src/TortoiseProc/LogDlg.cpp:
```diff
@@ -1285,6 +1285,92 @@ void CLogDlg::OnContextMenu(CWnd* pWnd, CPoint point)
+			m_LogList.m_Filter.m_NumberOfLogsScale = -1;
+		else if (cmd >= offset && cmd < cnt)
+		{
+				CString strDate = m_FromDateHistory.GetEntry(cmd - offset + 1);
```
oops~ tab problem

YueLinHo commented on src/Git/Git.cpp:
```diff
@@ -988,9 +985,47 @@ CString CGit::GetLogCmd(const CString &range, const CTGitPath *path, int count,
+			Filter->m_From = (__time64_t)(_tcstoui64(str.GetBuffer(0), nullptr, 10));
+			str.ReleaseBuffer();
+			if (Filter->m_From > 0)
+				st1.Format(_T(" --max-age=%I64u "), Filter->m_From);
```
Not sure line 995~999.
Should they move outside of this function?
(Put them in GitLogListBase.cpp?)

YueLinHo commented on src/Git/Git.h:
```diff
@@ -35,11 +35,32 @@ struct git_repository;
+		m_NumberOfLogs = (DWORD)CRegDWORD(_T("Software\\TortoiseGit\\NumberOfLogs"), 0);
+		m_NumberOfYears = (DWORD)CRegDWORD(_T("Software\\TortoiseGit\\NumberOfYears"), 0);
+		m_NumberOfMonths = (DWORD)CRegDWORD(_T("Software\\TortoiseGit\\NumberOfMonths"), 0);
+		m_NumberOfWeeks = (DWORD)CRegDWORD(_T("Software\\TortoiseGit\\NumberOfWeeks"), 0);
```
Put these regs under ```Software\TortoiseGit\LogDialog\```, instead of ```Software\TortoiseGit\```???

csware replied:
Yeah.
Also, do we need all these different keys?

YueLinHo replied:
I think we need.
1. Can switch different range of time quickly in Log Message dialog. (1 Year, 6 months(half year), 13 weeks(3 months)).

2. For first scale - say last commits, the "number of" setting value could be 10000. If we just have one "number of" setting, once  user switches to "year" scale, it will be 10000 year, user will need to modify that setting value. mum... not bad, not good.

3. The bad thing I could count is too many code added. :P

P.S. "number of" setting value for "last selected date".

YueLinHo replied:
@csware 
I found some code in D:\TortoiseGit\src\TortoiseProc\Commands\LogCommand.cpp, 
probably from TSvn:
```
bool LogCommand::Execute()
{
	...
	CString val = parser.GetVal(_T("limit"));
	int limit = _tstoi(val);
	...
	if (limit == 0)
	{
		CRegDWORD reg = CRegDWORD(_T("Software\\TortoiseGit\\NumberOfLogs"), 100);
		limit = (int)(LONG)reg;
	}
	...
	CLogDlg dlg;
	theApp.m_pMainWnd = &dlg;
	dlg.SetParams(orgCmdLinePath, cmdLinePath, rev, range, limit);
```


3 questions:

1. The ```limit``` option is there, but limit(```CLogDlg::m_limit```) is not used any more. Fix it or drop it?
2. If fix limit option, use ```Software\TortoiseGit\LogDialog\``` or keep ```Software\TortoiseGit\```???
3. If fix limit option, add one more ```scale``` option for LogCommand?

My thoughts for those 3 questions:

1. fix limit option, because it can be done in this PR.
2. keep ```Software\TortoiseGit\``` for upstream compatibility reason.
3. no need scale option. Also for upstream compatibility reason. Once the limit option was given, uses the last commits of scale. (But, I can imagine some users may fire a issue to add it in the future.)

YueLinHo commented on src/TortoiseProc/LogDataVector.cpp:
```diff
@@ -64,7 +64,10 @@ int CLogDataVector::ParserFromLog(CTGitPath *path, int count, int infomask, CStr
+	CFilterData filter;
+	filter.m_NumberOfLogs = count;
+	filter.m_NumberOfLogsScale = CFilterData::SHOW_LAST_N_COMMITS;
+	CString cmd = g_Git.GetLogCmd(gitrange, path, infomask, true, &filter);
```
There is only one call of this function (CLogDataVector::ParserFromLog()) which pass the parameter count = 2, otherwise count = -1.

I tested that call (count = 2) manually.
I am considering add some auto tests.

YueLinHo commented on src/TortoiseProc/LogDlg.cpp:
```diff
@@ -488,16 +493,10 @@ LRESULT CLogDlg::OnLogListLoading(WPARAM wParam, LPARAM /*lParam*/)
+		CTime begin, end;
+		m_LogList.GetTimeRange(begin, end);
+		m_DateFrom.SetTime(&begin);
+		m_DateTo.SetTime(&end);
```
Not perfect here.
It always updates the date to date control for now.

csware commented:
The context menu on the from date picker isn't optimal as it's not intuitive to look there.

csware commented:
There seems to be a timezone issue here for the filter when using a stored one.

YueLinHo commented:
> The context menu on the from date picker isn't optimal as it's not intuitive to look there.

agree. that's a problem. :(
Once this PR is merged, no one will know there is a new context menu available.

How about convert FROM and TO date controls to a TEXT control?
Another setting dialog will be popup after clicking that TEXT control.
There are FROM and TO date controls, 1 Year radio button, 6 months radio button, and so on.
OR...
Implement [that comment](https://code.google.com/p/tortoisegit/issues/detail?id=2078#c5), a ghosted line of text.

any better/other idea, currently?
@ch3cooli do you?

> There seems to be a timezone issue here for the filter when using a stored one.

Yap. I did not consider that. Will study how to fix it. Thank you.

YueLinHo commented on src/TortoiseProc/LogDlg.cpp:
```diff
@@ -1285,6 +1285,92 @@ void CLogDlg::OnContextMenu(CWnd* pWnd, CPoint point)
+		CString last = CString(MAKEINTRESOURCE(IDS_LAST));
+
+		CRegDWORD logs = CRegDWORD(_T("Software\\TortoiseGit\\NumberOfLogs"), 0);
+		item.Format(_T("%s %d %s"), last, (DWORD)logs, CString(MAKEINTRESOURCE(IDS_COMMITS)));
```
These 2 lines(1301, 1302) should be:
```
		item.Format(_T("%s %d %s"), last, m_LogList.m_Filter.m_NumberOfLogs, CString(MAKEINTRESOURCE(IDS_COMMITS)));
```

(also for following 3 items.)

YueLinHo commented:
@csware 
Keep the original commits at branch [i1003-0](https://github.com/YueLinHo/TortoiseGit/tree/i1003-0)
Rebase original branch onto last master(9d9617bd45f8cc2b95d1562c8886cc1103894a29) and give them a branch [i1003-1](https://github.com/YueLinHo/TortoiseGit/tree/i1003-1), it's just for comparing the updated PR easily.

So, the updated PR:

1. Move some reg keys reading code outside of CGit::GetLogCmd() function and CFilterData class.
(Move to constructor of CGitLogListBase)

2. Put those reg keys under ```\Software\TortoiseGit\LogDialog\```

3. Added CGitLogListBase::ConvertStrUtcDateToLocalTime() for time zone problem.

4. The "Number of" for last selected data is always 1.

5. Change ```CFilterData::SHOW_LAST_SELECTED_DATE``` 
to ..........```CFilterData::SHOW_SELECTED_DATE```. (Remove ```_LAST```)

6. Fix ```limit``` option of log command. Doc is also updated.

7. The tool tip of "number of " is changed.

8. Renamed the FROM date history reg key. (LogDlg_FromDate -> LogDlg_FromDate**s**)

9. Changed the rule of updating FROM date control after loading all logs.

10. Code changed for the context menu of FROM date control.

11. a little bug in CLogDlg::OnDtnDatetimechangeDatefrom(). (m_NumberOfLogsScale is not set)

12.  Fix tab problem

13. Squash your commit (but loss all info from that commit.)

YueLinHo commented:
@csware 
Since PR #182 was merged, I  imagine some conflict will be here.
So, rebase it and resolve some resource conflicts.

ch3cooli commented:
Is it possible to change background color of the datetimepicker so that user knows he has applied filter rathan "All" ?

ch3cooli commented:
Is it possible to change background color of the datetimepicker so that user knows he has applied filter rathan "All" ? Otherwise the user may forget to reset the filter and think the result is strange.
This is because in the old versions, date filters were not "remembered".

YueLinHo commented:
> Is it possible to change background color of the datetimepicker so that user knows he has applied filter rathan "All" ?
Otherwise the user may forget to reset the filter and think the result is strange.
This is because in the old versions, date filters were not "remembered".

@ch3cooli
**Totally agree.**
I often forget the default filter is applied when I used this feature these days.
Especially, I typed some keyword in filter next by TO date control.
It should find something out, but got nothing, then I realized that the FROM date filter is applied. -_-!
Therefor, the idea of background color is really a nice one. ^^d
I will try to figure it out. **Which color would be better? red, green, blue, or yellow?**

A note: a user's suggestion: https://code.google.com/p/tortoisegit/issues/detail?id=2078#c5
But, I don't think it is better than the idea of background color.

BTW, I think the context menu of FROM date should show which filter is applied, too.

csware commented:
I suggest not to use red and green

YueLinHo commented:
If using background, the solution might be derive from CDateTimeCtrl and handle WM_ERASEBKGND.
Do you like "derive"?

I got another Idea is that using balloon tip, but do not figure it out.
The ShowEditBalloon() only works for Edit control.
Is balloon tip a possible way?

Settings dialog is also not good.
The code breaks the simpleness.
Do you want to introduce one more dialog "General/Dialog 4"?
(I mean IDD_SETTINGSDIALOGS4​.)
And use 4 radio buttons for 4 kinds of log scale?

Need your comments!

csware commented:
Deriving is a common concept in OO. - Maybe it's also possible to extend the built in drop down stuff - then we do not need our own context menu.

A balloon tip isn't always visible?!

YueLinHo commented:
OK. Trying derive. :)

about balloon tip, originally, I plan to handle it in https://github.com/TortoiseGit/TortoiseGit/blob/master/src/TortoiseProc/LogDlg.cpp#L438
In GITLOG_START and GITLOG_END if filter is applied.
And maybe also show it when scroll down to the bottom of log.

BUT! Must have a balloon tip component for CDateTimeCtrl first. :P

YueLinHo commented:
Rebase and add one more commit to draw "From:" text background with the shell color of our tortoise icon.
![draw_from](https://cloud.githubusercontent.com/assets/5831804/8821924/df2d9a76-3093-11e5-9893-d594bdd2e8ec.png)

mum... a little ugly. ^^!

csware commented:
That could be a way to go...

YueLinHo commented:
Are you going to merge it? :-)

ch3cooli commented:
The background color indication is acceptable to me.

csware commented:
I'm rebasing this. Just some comments about stuiff I'm fixing.

csware commented on src/Git/Git.h:
```diff
@@ -36,11 +36,32 @@ struct git_repository;
 	{
 		m_From=m_To=-1;
 		m_IsRegex=1;
+		m_NumberOfLogsScale = -1;
```
`m_NumberOfLogsScale` is unsigned, thus, no -1

csware commented on src/Resources/TortoiseProcENG.rc:
```diff
@@ -207,7 +207,7 @@ CAPTION "Log Messages"
 BEGIN
     LTEXT           "Branch",IDC_STATIC_REF,7,9,90,8,SS_NOTIFY | SS_ENDELLIPSIS
-    LTEXT           "From:",IDC_FROMLABEL,98,9,30,8
+    LTEXT           "  From:",IDC_FROMLABEL,97,5,84,19,SS_CENTERIMAGE
```
Is there another way than adding two spaces?

YueLinHo replied:
Perhaps OnPaint()

csware commented on src/TortoiseProc/LogDlg.cpp:
```diff
@@ -431,10 +455,35 @@ BOOL CLogDlg::OnInitDialog()
+{
+	__super::OnDestroy();
+
+	m_Brush.DeleteObject();
```
Why not in the destructor?

YueLinHo replied:
1. Old habits die hard!
Create in OnInitDialog(), destory in OnDestory()

2. I am not sure if it is suitable to delete GDI object in destructor.
Is it OK?

YueLinHo replied:
I will drop this, because CBrush do DeleteObject() in its destructor.
See C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\atlmfc\include\afxwin1.inl Line 75

csware commented on src/TortoiseProc/Settings/SettingsAdvanced.cpp:
```diff
@@ -86,6 +86,10 @@ CSettingsAdvanced::CSettingsAdvanced()
 	settings[i].type	= CSettingsAdvanced::SettingTypeNumber;
 	settings[i++].def.l	= 5;
 
+	settings[i].sName   = L"MaxFromDateHistoryItems";
```
Did you want to drop this?

YueLinHo replied:
Drop is OK. :)

csware commented on src/Resources/TortoiseProcENG.rc:
```diff
@@ -690,39 +690,43 @@ BEGIN
-    COMBOBOX        IDC_FONTNAMES,107,17,136,90,CBS_DROPDOWNLIST | CBS_OWNERDRAWVARIABLE | CBS_SORT | CBS_HASSTRINGS | WS_VSCROLL | WS_TABSTOP
-    COMBOBOX        IDC_FONTSIZES,253,17,31,47,CBS_DROPDOWN | WS_VSCROLL | WS_TABSTOP
+    GROUPBOX        "Log messages",IDC_STATIC,7,7,287,240
+    LTEXT           "Default limitation of log messages:",IDC_STATIC,14,20,150,8,SS_CENTERIMAGE
```
What about adding a selector for:
* No limitation
* remember last date
* set custom limit

YueLinHo replied:
Good idea! ^_^

YueLinHo replied:
But, what is "set custom limit"?

csware replied:
User specified/specific

YueLinHo replied:
still do not understand. How?
Could you please say more?

csware replied:
I don't like the zero in the field for no limit.

YueLinHo replied:
> I don't like the zero in the field for no limit.

That's OK, I add the "No limitation" as you suggested.

YueLinHo replied:
> User specified/specific

Do you mean?
memorize the last selected item when closing dialog.
use the last selected item, ex: last 1 year, when re-open dialog.

csware replied:
Adjust the settings optics:
1) Allow user to choose: No limit, remeber last selected date limit, used custom limit and then on the next line the user can specify it.

YueLinHo replied:
oops, I just implement some.
see https://github.com/YueLinHo/TortoiseGit/tree/i1003-test

csware commented on src/TortoiseProc/LogDlg.cpp:
```diff
@@ -208,7 +208,7 @@ enum JumpType
 };
 
-void CLogDlg::SetParams(const CTGitPath& orgPath, const CTGitPath& path, CString hightlightRevision, CString range, int limit)
+void CLogDlg::SetParams(const CTGitPath& orgPath, const CTGitPath& path, CString hightlightRevision, CString range, int limit, int limitScale/*=0*/)
```
`limit` and `limitScale` as DWORD?

YueLinHo replied:
Will change them to DWORD.

csware commented on src/Git/Git.h:
```diff
@@ -36,11 +36,32 @@ struct git_repository;
+
+	enum
+	{
+		SHOW_LAST_N_COMMITS = 0,
```
Add `SHOW_NO_LIMIT`?

YueLinHo replied:
GOOD! :)
Thanks for that.

csware commented on src/Resources/TortoiseProcENG.rc:
```diff
@@ -690,39 +690,43 @@ BEGIN
+    LTEXT           "Default limitation of log messages:",IDC_STATIC,14,20,150,8,SS_CENTERIMAGE
+    LTEXT           "Last",IDC_STATIC,170,20,26,8,SS_CENTERIMAGE
+    EDITTEXT        IDC_DEFAULT_NUMBER_OF,199,18,28,13,ES_AUTOHSCROLL,WS_EX_RIGHT
+    COMBOBOX        IDC_DEFAULT_SCALE,229,18,55,47,CBS_DROPDOWNLIST | WS_VSCROLL | WS_TABSTOP
```
The combobox must be wider for translations

csware commented on src/TortoiseProc/LogDlg.cpp:
```diff
@@ -1293,6 +1339,75 @@ void CLogDlg::OnContextMenu(CWnd* pWnd, CPoint point)
+			return;
+
+		int cnt = 0;
+		CTime time;
```
Declare in smalles scope

csware commented:
See https://download.tortoisegit.org/tmp/notes.patch

csware commented on src/TortoiseProc/GitLogListBase.cpp:
```diff
@@ -95,8 +95,16 @@ CGitLogListBase::CGitLogListBase():CHintListCtrl()
-	m_From = -1;
-	m_To = -1;
+	m_Filter.m_NumberOfLogsScale = (DWORD)CRegDWORD(_T("Software\\TortoiseGit\\LogDialog\\NumberOfLogsScale"), -1);
+	m_Filter.m_NumberOfLogs = (DWORD)CRegDWORD(_T("Software\\TortoiseGit\\LogDialog\\NumberOfLogs"), 0);
```
DO we need all these different limits? Isn't just one "nNumber" sufficient?

YueLinHo replied:
It shares between commits/years/months/weeks.
the value coulde be 1000/1/6/13
It will be more friendly for user. (but more code.)
I you want one "m_nNumber", I will do it. :)

csware replied:
Have you used these different numbers so far?

YueLinHo replied:
No, I use last 1000 commits and no limitation.
Actually, I prefer to drop "Last N years/months/weeks", but user request those.

csware replied:
Did the user requested to have those at the same this? E.g. last 3 weeks, last 5 months, 6 years to select from? - I thought our users just wanted to be able to set a concrete limit "5 years" OR "3 months" instead of last 400 commits.

csware replied:
Or is it better to arrange it differently: A checkbox for "Remeber last selected from filter" and then allow to user to optionally add a default limitation for x commits/days/weeks/months/years.

YueLinHo replied:
if those values are dropped, the "from" context menu will have the same value on commits/years/months/weeks.
It leads some context menu items are also needed to be dropped, and leave one item there.
maybe "Last 1000 commits", or "Last 1 year". (I accept this. :) )

YueLinHo replied:
> Remeber last selected from filte

I guess I implemet it [here](https://github.com/YueLinHo/TortoiseGit/tree/i1003-test). Right?

csware replied:
About your `i1003-test` branch:
* I like the settings dialog
* Why is the From date surrounded with a border when no limitation is set?
* Do we need all those options in the "Last x .." entries in the context menu?
* Maybe add a checkbox entry in the context menu for "remember selected date on close"
* It does not compile ;)

YueLinHo replied:
​​
>About your i1003-test branch:
I like the settings dialog

Thank you. :)

> There is no way for making the log dialog "remembering" the last selected date on close

Fixed. :)

>Why is the From date surrounded with a border when no limitation is set?

Obviously, my fault, no "if(){}" in OnPaint().
Fixed with no border there.

>Do we need all those options in the "Last x .." entries in the context menu?

Dropped them.

>Maybe add a checkbox entry in the context menu for "remember selected date on close"

Added.

>It does not compile ;) (just 2 minor c&p errors)

Oops. Also fixed.

----
All of all, thank you for pointing these out. appreciate. ^_^

csware commented:
doc\source\en\TortoiseGit\tgit_app_automation.xml should also updated.

YueLinHo commented:
> doc\source\en\TortoiseGit\tgit_app_automation.xml

OK. will do that. :)

YueLinHo commented:
Rebase.
Dropped those var for "Last N" options, only one var now.
Dropped "Last N" context menu items, too.

One thing I am not sure is doc.

csware commented:
"Remeber last selected date" does not work, also default limitation to x commits does not work.

csware commented on Languages/Tortoise.pot:
```diff
@@ -5803,10 +5811,34 @@ msgstr ""
+msgstr ""
+
+#. Resource IDs: (1579)
+msgid "Last selected item"
```
What's this?

YueLinHo replied:
choose context menu item "All" last time, show all commits next time.
choose some date last time, show that date next time.

csware replied:
Ah, ok I see. Don't think that this is clear to all users.

YueLinHo replied:
Agree with "Not clear". :)

I will drop it. or you do it for me?
I am going to sleep. see u. :)

csware replied:
I'm thinking of a better wording/solution...

YueLinHo commented:
> "Remeber last selected date" does not work, also default limitation to x commits does not work.

Last commit is broken.

YueLinHo commented:
@csware fixed last commit.
Also updated the tooltips for those new controls in settings dialog.

csware commented:
What's the difference between "Last selected item" and "Last selected date"?

csware commented:
Merged with modifications.
History stuff should be added separately if we still want it (I remember from our conversations that you never used those, thus, I dropped it for now).

AndrewLane commented:
Woohoo!  Thanks for all your work on this!

YueLinHo commented:
Thanks for modification for me and thanks for merge. ^_^
